pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/UnAckedMessageTracker.java

```diff
@@ -88,7 +90,7 @@ public void start(PulsarClientImpl client, ConsumerBase consumerBase, long ackTi
             public void run(Timeout t) throws Exception {
                 if (isAckTimeout()) {
                     log.warn("[{}] {} messages have timed-out", consumerBase, oldOpenSet.size());
-                    List<MessageIdImpl> messageIds = new ArrayList<>();
+                    Set<MessageIdImpl> messageIds = new HashSet<>();
```

Comment:
I think introducing ```HashSet``` is not a good idea as redelivery entries will not be in sequence in this case and instead of sequentially, broker will try to read random entries from bookkeeper which we should avoid.